src/SprykerSdk/Zed/ComposerReplace/Business/ComposerReplaceFacade.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Zed\ComposerReplace\Business;
+
+use Generated\Shared\Transfer\ReplaceUpdateResultTransfer;
+use Generated\Shared\Transfer\ReplaceValidationResultTransfer;
+use Spryker\Zed\Kernel\Business\AbstractFacade;
+
+/**
+ * @api
+ *
+ * @method \SprykerSdk\Zed\ComposerReplace\Business\ComposerReplaceBusinessFactory getFactory()
+ */
+class ComposerReplaceFacade extends AbstractFacade implements ComposerReplaceFacadeInterface
+{
+    /**
+     * {@inheritdoc}
+     *
+     * @api
+     *
+     * @return \Generated\Shared\Transfer\ReplaceValidationResultTransfer
+     */
+    public function validate(): ReplaceValidationResultTransfer
```

Comment:
Can you please refactor this to not use two completely different facade methods for this, but the same one with the only difference of "dry run" true?
This here creates the possible unDRY issue of validate vs update to become out of sync, and duplicates too much logic.
Ideally, the dry-run executes the same path just without the last part of modifying.

Is this someone possible to achieve? Either via same facade or through underlying business logic then at least.
We had the same issue already in the other composer tool that also behaves slightly different easily based on modifying one business class, but forgetting to replicate the same on the other side, resulting in a totally weird result.

Reply:
Could you point me to the code duplications you see in this PR, please? There is an abstract class in place which holds all from both classes used code, the rest is individual code.

Using one method as you explained would be a violation of the Single responsibility principle.

Reply:
It is in this case not the duplication, but completely different paths for a while.

validateComposerJsonReplace() uses
- validateMissingComposerPackages()
- validateObsoleteComposerPackages()

in isolation compared to

updateComposerJsonReplace() which uses
- addComposerPackageToReplace()
- removeComposerPackageFromReplace()

How can you make sure future changes in one of the two are reflected properly to be always in sync.

For me, a proper way of doing this is to ensure the collecting is the same for both, and the output is also the same, the only difference is the dry-run would not write to file (simulate changes only and display on CLI).

This can usually be done two ways:
- updater internally uses validator(), with the addition of writing
- there is only updater, but with dry-run flag or different entry point method that adds a protected write() at the end (which basically is the same as option 1, only inside the same file somewhat).

Reply:
This is basically an antipattern of SOLID, as it takes it to an extreme that creates bugs.
SOLID must not be taken so seriously to create these issues in coding.
In these cases, single responsibility also means "same responsibility", not "isolation" and thus un-DRY timebomb.

We should formalize this as important coding pattern for the CLI dry-run pattern of fixer scripts.